### PR TITLE
HOLO-1394: Add NFT asset class

### PR DESCRIPTION
**Pull request by feliperodriguess**
## Describe Changes

Changes made to the package: **sdk**.

I made it better by doing ...

- Add NFT asset class
- Add validation
- Add error mapping/handling
- Add `IsNotMinted()` decorator
- Add tests

Screenshot 2024-03-05 at 20 46 14

## Checklist before requesting a review

- [x] I have performed a self-review of my code
- [x] Code styles have been enforced
- [x] All tests are passing

### CODE REVIEW

File: packages/sdk/src/assets/nft.ts [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+  getOwner() {
+    return this.owner
+  }
+
+  getFile() {
+    return this.file
+  }
+
+  getTokenId() {
+    if (!this.chainId || !this.tokenId) throw new Error(NOT_MINTED_NFT_ERROR_MESSAGE)
```
Comment (sogoiii) [inline packages/sdk/src/assets/nft.ts:62]: @feliperodriguess oh nice! i like this error. Just FYI we have our own Error object `HolographError` that we can use instead.

File: packages/sdk/src/assets/nft.ts [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+      part: {
+        chainId: parseInt(chainIdHex, 16).toString(),
+        decimal: parseInt(tokenNumberHex, 16).toString(),
+        hex: removeLeftZeroes(tokenNumberHex),
+      },
+    }
+  }
+
+  setMetadata(metadata: HolographNFTMetadata) {
+    if (this.isMinted) throw new Error(UPDATE_MINTED_NFT_ERROR_MESSAGE)
```
Comment (sogoiii) [inline packages/sdk/src/assets/nft.ts:79]: @feliperodriguess would it be worth it to make a decorator for the minted state check?

File: packages/sdk/src/assets/nft.ts [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    this.metadata.name = name
+  }
+
+  setDescription(description: string) {
+    if (this.isMinted) throw new Error(UPDATE_MINTED_NFT_ERROR_MESSAGE)
+    this.metadata.description = description
+  }
+
+  setCreator(creator: string) {
+    if (this.isMinted) throw new Error(UPDATE_MINTED_NFT_ERROR_MESSAGE)
```
Comment (sogoiii) [inline packages/sdk/src/assets/nft.ts:94]: @feliperodriguess i know the other keys don't have validation. But we should make sure that the input is an address string.

File: packages/sdk/src/assets/nft.ts [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+    this.metadata.creator = creator
+  }
+
+  setAttributes(attributes: {[key: string]: string}) {
+    if (this.isMinted) throw new Error(UPDATE_MINTED_NFT_ERROR_MESSAGE)
+    this.metadata.attributes = attributes
+  }
+
+  setFile(file: string) {
+    if (this.isMinted) throw new Error(UPDATE_MINTED_NFT_ERROR_MESSAGE)
```
Comment (sogoiii) [inline packages/sdk/src/assets/nft.ts:104]: @feliperodriguess Lets make sure that the input string is either `htttps://` or `ipfs://` urls.

File: packages/sdk/src/assets/nft.ts [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+  // TODO: Remove the following methods soon
+  setTokenId(tokenId: string) {
+    this.tokenId = tokenId
+  }
+
+  setChainId(chainId: string | number) {
+    this.chainId = chainId
+  }
+
+  toggleIsMinted() {
```
Comment (sogoiii) [inline packages/sdk/src/assets/nft.ts:124]: @feliperodriguess lets label this as a helper function: "Toggle function used for easy testing while in development. Remove before public release"

File: packages/sdk/src/tests/assets/nft.test.ts [hunk 6]
```diff
@@ -0,0 +1,246 @@
+import {beforeEach, describe, expect, it} from 'vitest'
+
+import {NFT, NOT_MINTED_NFT_ERROR_MESSAGE, UPDATE_MINTED_NFT_ERROR_MESSAGE} from '../../assets/nft'
+
+const expectedTokenId = {
```
Comment (sogoiii) [inline packages/sdk/src/tests/assets/nft.test.ts:5]: @feliperodriguess From a previous conversations, i asked to change the keys to the following shape:

```
 {
  hex,
  decimal,
  part: { chainId, tokenNumber }
}
```

**Review by sogoiii — CHANGES_REQUESTED** [review]

File: packages/sdk/src/assets/nft.ts (hunk 5 above)
Comment (feliperodriguess) [reply packages/sdk/src/assets/nft.ts:127]: ✅

File: packages/sdk/src/assets/nft.ts (hunk 4 above)
Comment (feliperodriguess) [reply packages/sdk/src/assets/nft.ts]: ✅

File: packages/sdk/src/assets/nft.ts (hunk 3 above)
Comment (feliperodriguess) [reply packages/sdk/src/assets/nft.ts]: Creator is not an address, it's the name of the creator person/company

The `owner` is the one that is an address, I created a validator for that

File: packages/sdk/src/assets/nft.ts (hunk 2 above)
Comment (feliperodriguess) [reply packages/sdk/src/assets/nft.ts]: ✅

File: packages/sdk/src/assets/nft.ts (hunk 1 above)
Comment (feliperodriguess) [reply packages/sdk/src/assets/nft.ts]: ✅

File: packages/sdk/src/tests/assets/nft.test.ts (hunk 6 above)
Comment (feliperodriguess) [reply packages/sdk/src/tests/assets/nft.test.ts:7]: ✅

**Review by sogoiii — CHANGES_REQUESTED** [review]
After these changes we can move on. 

1. rename the `/assets/utils.ts` to `/assets/nft.validation.ts` Or just put the validation definition in the same file as `/assets/nft.ts`. the name `utils.ts` is not very helpful. 
2. rename `NotMintedOnly` to `isNotMinted`. The original name is a little awkward.

**Comment by feliperodriguess** [comment]
> After these changes we can move on.
> 
> 1. rename the `/assets/utils.ts` to `/assets/nft.validation.ts` Or just put the validation definition in the same file as `/assets/nft.ts`. the name `utils.ts` is not very helpful.
> 2. rename `NotMintedOnly` to `isNotMinted`. The original name is a little awkward.

Done

File: packages/sdk/src/assets/nft.ts [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+  getOwner() {
+    return this.owner
+  }
+
+  getFile() {
+    return this.file
+  }
+
+  getTokenId() {
+    if (!this.chainId || !this.tokenId) throw new NotMintedNftError(this.getTokenId.name)
```
Comment (BravoNatalie) [inline packages/sdk/src/assets/nft.ts:52]: I believe that it's better to keep the same pattern for this type of verification. This was a suggestion made by @alexanderattar in the CLI.

```ts
if(this.chainId !== undefined || this.token !== undefined) {
   throw new NotMintedNftError(this.getTokenId.name)
}
```

**Review by BravoNatalie — APPROVED** [review]

File: packages/sdk/src/assets/nft.ts (hunk 7 above)
Comment (feliperodriguess) [reply packages/sdk/src/assets/nft.ts:52]: I think this is not what I was expecting. I wanna throw an error either if `this.chainId` or `this.tokenId` got any `falsy` value, not only `undefined`

**Pull request merged**